packages/worker/src/workflow/vm-shared.ts

```diff
@@ -270,78 +271,122 @@ export const globalHandlers = new GlobalHandlers();
+    // Update the activator's state in preparation for a non-query activation.
+    // This is done early, so that we can then rely on the activator while processing the activation.
+    if (activation.timestamp == null)
+      throw new TypeError('Expected activation.timestamp to be set for non-query activation');
+    this.activator.now = tsToMs(activation.timestamp);
+    this.activator.mutateWorkflowInfo((info) => ({
+      ...info,
+      historyLength: activation.historyLength as number,
+      // Exact truncation for multi-petabyte histories
+      // historySize === 0 means WFT was generated by pre-1.20.0 server, and the history size is unknown
+      historySize: activation.historySizeBytes?.toNumber() ?? 0,
+      continueAsNewSuggested: activation.continueAsNewSuggested ?? false,
+      currentBuildId: activation.buildIdForCurrentTask ?? undefined,
+      unsafe: {
+        ...info.unsafe,
+        isReplaying: activation.isReplaying ?? false,
+      },
+    }));
+    this.activator.addKnownFlags(activation.availableInternalFlags ?? []);
+
     const [patches, nonPatches] = partition(activation.jobs, ({ notifyHasPatch }) => notifyHasPatch != null);
-    const [signals, nonSignals] = partition(nonPatches, ({ signalWorkflow }) => signalWorkflow != null);
-    const [queries, rest] = partition(nonSignals, ({ queryWorkflow }) => queryWorkflow != null);
-    let batchIndex = 0;
+    for (const { notifyHasPatch } of patches) {
+      if (notifyHasPatch == null) throw new TypeError('Expected notifyHasPatch to be set');
+      this.activator.notifyHasPatch(notifyHasPatch);
+    }
 
-    // Loop and invoke each batch and wait for microtasks to complete.
-    // This is done outside of the isolate because when we used isolated-vm we couldn't wait for microtasks from inside the isolate, not relevant anymore.
-    for (const jobs of [patches, signals, rest, queries]) {
-      if (jobs.length === 0) {
-        continue;
-      }
+    const hasSignals = activation.jobs.some(({ signalWorkflow }) => signalWorkflow != null);
+    const doSingleBatch = !hasSignals || this.activator.hasFlag(SdkFlags.ProcessWorkflowActivationJobsAsSingleBatch);
+
+    const batches: coresdk.workflow_activation.IWorkflowActivationJob[][] = doSingleBatch
+      ? [nonPatches]
+      : partition(
+          nonPatches,
+          // Move signals to a first batch; all the rest goes in a second batch.
+          ({ signalWorkflow }) => signalWorkflow != null
+        );
+
+    // Loop and invoke each batch, waiting for microtasks to complete after each batch.
+    let batchIndex = 0;
+    for (const jobs of batches) {
+      if (jobs.length === 0) continue;
       this.workflowModule.activate(
         coresdk.workflow_activation.WorkflowActivation.fromObject({ ...activation, jobs }),
         batchIndex++
       );
-      if (internals.shouldUnblockConditions(jobs[0])) {
-        this.tryUnblockConditions();
-      }
+      this.tryUnblockConditionsAndMicrotasks();
```

Comment:
Even though it would take up more lines, I'd kinda prefer for this to read as

```
if doSingleBatch {
    // new code with no loop
} else {
    oldMultiBatchProcessing();
}
```

Which would allow the new path to be free of some of the clutter of the old path. Up to you though.

Reply:
Good point.